File: app/src/main/java/com/lipisoft/toyshark/transport/tcp/TCPHeader.java
/*
 *  Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.lipisoft.toyshark.transport.tcp;

import android.support.annotation.Nullable;

import com.lipisoft.toyshark.transport.ITransportHeader;

/**
 * data structure for TCP Header
 *
 * @author Borey Sao
 * Date: May 8, 2014
 */
public class TCPHeader implements ITransportHeader {
    private int sourcePort;
    private int destinationPort;
    private long sequenceNumber; // 32 bits
    private int dataOffset;
    private int tcpFlags;
    private boolean isNS;
    private boolean isCWR = false;
    private boolean isECE = false;
    private boolean isSYN = false;
    private boolean isACK = false;
    private boolean isFIN = false;
    private boolean isRST = false;
    private boolean isPSH = false;
    private boolean isURG = false;
    private int windowSize;
    private int checksum;
    private int urgentPointer;
    @Nullable
    private byte[] options;
    private long ackNumber; // 32 bits

    // vars below need to be set via setters when copy
    private int maxSegmentSize = 0;
    private int windowScale = 0;
    private boolean isSelectiveAckPermitted = false;
    private int timeStampSender = 0;
    private int timeStampReplyTo = 0;

    TCPHeader(int sourcePort, int destinationPort, long sequenceNumber, long ackNumber,
              int dataOffset, boolean isns, int tcpFlags,
              int windowSize, int checksum, int urgentPointer) {
        this.sourcePort = sourcePort;
        this.destinationPort = destinationPort;
        this.sequenceNumber = sequenceNumber;
        this.dataOffset = dataOffset;
        this.isNS = isns;
        this.tcpFlags = tcpFlags;
        this.windowSize = windowSize;
        this.checksum = checksum;
        this.urgentPointer = urgentPointer;
        this.ackNumber = ackNumber;
        setFlagBits();
    }

    private void setFlagBits() {
        isFIN = (tcpFlags & 0x01) > 0;
        isSYN = (tcpFlags & 0x02) > 0;
        isRST = (tcpFlags & 0x04) > 0;
        isPSH = (tcpFlags & 0x08) > 0;
        isACK = (tcpFlags & 0x10) > 0;
        isURG = (tcpFlags & 0x20) > 0;
        isECE = (tcpFlags & 0x40) > 0;
        isCWR = (tcpFlags & 0x80) > 0;
    }

    public boolean isNS() {
        return isNS;
    }

    void setIsNS(boolean isNS) {
        this.isNS = isNS;
    }

    public boolean isCWR() {
        return isCWR;
    }

    void setIsCWR(boolean isCWR) {
        this.isCWR = isCWR;
        if (isCWR) {
            this.tcpFlags |= 0x80;
        } else {
            this.tcpFlags &= 0x7F;
        }
    }

    public boolean isECE() {
        return isECE;
    }

    void setIsECE(boolean isECE) {
        this.isECE = isECE;
        if (isECE) {
            this.tcpFlags |= 0x40;
        } else {
            this.tcpFlags &= 0xBF;
        }
    }

    public boolean isSYN() {
        return isSYN;
    }

    void setIsSYN(boolean isSYN) {
        this.isSYN = isSYN;
        if (isSYN) {
            this.tcpFlags |= 0x02;
        } else {
            this.tcpFlags &= 0xFD;
        }
    }

    public boolean isACK() {
        return isACK;
    }

    void setIsACK(boolean isACK) {
        this.isACK = isACK;
        if (isACK) {
            this.tcpFlags |= 0x10;
        } else {
            this.tcpFlags &= 0xEF;
        }
    }

    public boolean isFIN() {
        return isFIN;
    }

    void setIsFIN(boolean isFIN) {
        this.isFIN = isFIN;
        if (isFIN) {
            this.tcpFlags |= 0x1;
        } else {
            this.tcpFlags &= 0xFE;
        }
    }

    public boolean isRST() {
        return isRST;
    }

    void setIsRST(boolean isRST) {
        this.isRST = isRST;
        if (isRST) {
            this.tcpFlags |= 0x04;
        } else {
            this.tcpFlags &= 0xFB;
        }
    }

    public boolean isPSH() {
        return isPSH;
    }

    void setIsPSH(boolean isPSH) {
        this.isPSH = isPSH;
        if (isPSH) {
            this.tcpFlags |= 0x08;
        } else {
            this.tcpFlags &= 0xF7;
        }
    }

    public boolean isURG() {
        return isURG;
    }

    void setIsURG(boolean isURG) {
        this.isURG = isURG;
        if (isURG) {
            this.tcpFlags |= 0x20;
        } else {
            this.tcpFlags &= 0xDF;
        }
    }

    public int getSourcePort() {
        return sourcePort;
    }

    void setSourcePort(int sourcePort) {
        this.sourcePort = sourcePort;
    }

    public int getDestinationPort() {
        return destinationPort;
    }

    void setDestinationPort(int destinationPort) {
        this.destinationPort = destinationPort;
    }

    public long getSequenceNumber() {
        return sequenceNumber;
    }

    void setSequenceNumber(long sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
    }

    public int getDataOffset() {
        return dataOffset;
    }

    public void setDataOffset(int dataOffset) {
        this.dataOffset = dataOffset;
    }

    int getTcpFlags() {
        return tcpFlags;
    }

    public void setTcpFlags(int tcpFlags) {
        this.tcpFlags = tcpFlags;
    }

    public int getWindowSize() {
        return windowSize;
    }

    void setWindowSize(int windowSize) {
        this.windowSize = windowSize;
    }

    public int getChecksum() {
        return checksum;
    }

    public void setChecksum(int checksum) {
        this.checksum = checksum;
    }

    public int getUrgentPointer() {
        return urgentPointer;
    }

    public void setUrgentPointer(int urgentPointer) {
        this.urgentPointer = urgentPointer;
    }

    @Nullable
    public byte[] getOptions() {
        return options;
    }

    void setOptions(@Nullable byte[] options) {
        this.options = options;
    }

    public long getAckNumber() {
        return ackNumber;
    }

    void setAckNumber(long ackNumber) {
        this.ackNumber = ackNumber;
    }

    /**
     * length of TCP Header including options length if available.
     *
     * @return int
     */
    public int getTCPHeaderLength() {
        return (dataOffset * 4);
    }

    public int getMaxSegmentSize() {
        return maxSegmentSize;
    }

    void setMaxSegmentSize(int maxSegmentSize) {
        this.maxSegmentSize = maxSegmentSize;
    }

    public int getWindowScale() {
        return windowScale;
    }

    void setWindowScale(int windowScale) {
        this.windowScale = windowScale;
    }

    boolean isSelectiveAckPermitted() {
        return isSelectiveAckPermitted;
    }

    void setSelectiveAckPermitted(boolean isSelectiveAckPermitted) {
        this.isSelectiveAckPermitted = isSelectiveAckPermitted;
    }

    public int getTimeStampSender() {
        return timeStampSender;
    }

    void setTimeStampSender(int timeStampSender) {
        this.timeStampSender = timeStampSender;
    }

    int getTimeStampReplyTo() {
        return timeStampReplyTo;
    }

    void setTimeStampReplyTo(int timeStampReplyTo) {
        this.timeStampReplyTo = timeStampReplyTo;
    }

}

File: app/src/main/java/com/lipisoft/toyshark/transport/tcp/TCPHeader.kt
/*
 *  Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.lipisoft.toyshark.transport.tcp

import com.lipisoft.toyshark.transport.ITransportHeader

/**
 * data structure for TCP Header
 *
 * @author Borey Sao
 * Date: May 8, 2014
 */
class TCPHeader (
        sourcePort: Int,
        destinationPort: Int,
        sequenceNumber: Long,
        ackNumber: Long,
        var dataOffset: Int,
        isNS: Boolean,
        tcpFlags: Int,
        windowSize: Int,
        var checksum: Int,
        var urgentPointer: Int) : ITransportHeader {

    override var sourcePort: Int = 0
    override var destinationPort: Int = 0

    var sequenceNumber: Long = 0 // 전송되는 데이터의 Byte 순서 번호, 랜덤한 숫자가 들어감

    var tcpFlags: Int = 0

    var isNS: Boolean = false // ECN-nonce 은폐 보호
    var isCWR = false // 혼잡 윈도 축소 플래그는 송식측 호스트에 의해 성정되는것
    var isECE = false // SYN=1 : TCP 가 명시적 혼잡 통지가 가능한 상태 / SYN=0 : IP 헤더 셋에 혼잡 경험 플래그가 설정된 패킷이 정상 수신
    var isSYN = false // 연결 요청 : TCP 에서 세션 성립시 가장 먼저 보내는 패킷, 임의의 시퀀스 번호를 보내어 설정
    var isACK = false // 응답 : 상대방으로부터 패킷을 받았다는 패킷 (보통 +1), ACK 응답으로 성공 실패를 판단
    var isFIN = false // 종료 : 세션 연결을 종료시킬때 사용, 더 이상 전송할 데이터가 없음
    var isRST = false // 강제 종료 : Reset 과정이며, 양방향에서 동시에 일어나는 중단 작업
    var isPSH = false // 밀어넣기 플래그 : 상호작용이 중요한 프로토콜인 경우 빠른 응답을 위해 7Layer에 전송
    var isURG = false // 긴급 : 헤더를 참조하여 어디~어디까지 우선 보내달라는 플래그

    var windowSize: Int = 0 // 이만큼 보낼 수 있다 (송신 윈도우), 이만큼 바등ㄹ 수 있다 (수신 윈도우) 최대:65535, 혼잡제어용
    var windowScale = 0

    var ackNumber: Long = 0 // 수신측에서 앞으로 받고자 하는 byte 순서 번호, 마지막으로 받은 데이터에 seq+1을 함
    var options: ByteArray? = null // 연결이 구성되는 동안 협상할 최대 세그먼트 옵션

    // vars below need to be set via setters when copy
    var maxSegmentSize = 0
    var isSelectiveAckPermitted = false
    var timeStampSender = 0
    var timeStampReplyTo = 0

    /**
     * length of TCP Header including options length if available.
     * @return int
     */
    val tcpHeaderLength: Int
        get() = dataOffset * 4

    init {
        this.sourcePort = sourcePort
        this.destinationPort = destinationPort
        this.sequenceNumber = sequenceNumber
        this.isNS = isNS
        this.tcpFlags = tcpFlags
        this.windowSize = windowSize
        this.ackNumber = ackNumber
        setFlagBits()
    }

    private fun setFlagBits() {
        isFIN = tcpFlags and 0x01 > 0
        isSYN = tcpFlags and 0x02 > 0
        isRST = tcpFlags and 0x04 > 0
        isPSH = tcpFlags and 0x08 > 0
        isACK = tcpFlags and 0x10 > 0
        isURG = tcpFlags and 0x20 > 0
        isECE = tcpFlags and 0x40 > 0
        isCWR = tcpFlags and 0x80 > 0
    }

    fun setIsCWR(isCWR: Boolean) {
        this.isCWR = isCWR
        if (isCWR) {
            this.tcpFlags = this.tcpFlags or 0x80
        } else {
            this.tcpFlags = this.tcpFlags and 0x7F
        }
    }

    fun setIsECE(isECE: Boolean) {
        this.isECE = isECE
        if (isECE) {
            this.tcpFlags = this.tcpFlags or 0x40
        } else {
            this.tcpFlags = this.tcpFlags and 0xBF
        }
    }

    fun setIsSYN(isSYN: Boolean) {
        this.isSYN = isSYN
        if (isSYN) {
            this.tcpFlags = this.tcpFlags or 0x02
        } else {
            this.tcpFlags = this.tcpFlags and 0xFD
        }
    }

    fun setIsACK(isACK: Boolean) {
        this.isACK = isACK
        if (isACK) {
            this.tcpFlags = this.tcpFlags or 0x10
        } else {
            this.tcpFlags = this.tcpFlags and 0xEF
        }
    }

    fun setIsFIN(isFIN: Boolean) {
        this.isFIN = isFIN
        if (isFIN) {
            this.tcpFlags = this.tcpFlags or 0x1
        } else {
            this.tcpFlags = this.tcpFlags and 0xFE
        }
    }

    fun setIsRST(isRST: Boolean) {
        this.isRST = isRST
        if (isRST) {
            this.tcpFlags = this.tcpFlags or 0x04
        } else {
            this.tcpFlags = this.tcpFlags and 0xFB
        }
    }

    fun setIsPSH(isPSH: Boolean) {
        this.isPSH = isPSH
        if (isPSH) {
            this.tcpFlags = this.tcpFlags or 0x08
        } else {
            this.tcpFlags = this.tcpFlags and 0xF7
        }
    }

    fun setIsURG(isURG: Boolean) {
        this.isURG = isURG
        if (isURG) {
            this.tcpFlags = this.tcpFlags or 0x20
        } else {
            this.tcpFlags = this.tcpFlags and 0xDF
        }
    }

}
